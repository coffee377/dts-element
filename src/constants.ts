import * as ts from 'typescript';
import {create_native_type, INativeType} from './types/native-type';

export enum ElementKind {
  GenericDeclaration,
  GenericType,
  NativeType,
  ArrayType,
  Parameter,
}

export const any_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)});
export const never_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NeverKeyword)});

export const null_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword)});
export const undefined_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.UndefinedKeyword)});
export const void_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)});

export const boolean_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword)});
export const number_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword)});
export const string_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword)});
export const symbol_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.SymbolKeyword)});
export const object_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ObjectKeyword)});

export const this_type = create_native_type({type: ts.createKeywordTypeNode(ts.SyntaxKind.ThisKeyword)});
