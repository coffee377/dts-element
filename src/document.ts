import {Element} from './element';
import {RootElement} from './sets';
import {trim_every_line} from './helpers/trim-every-line';
import {AnyElement} from './element';

export interface IDocumentRequiredParameters {
  elements: RootElement[];
}

// tslint:disable-next-line no-empty-interface
export interface IDocumentOptionalParameters {}

export class Document extends Element<IDocumentRequiredParameters, IDocumentOptionalParameters> {

  public get default_parameters(): IDocumentOptionalParameters {
    return {};
  }

  public _emit(_container: AnyElement): string {
    const {elements} = this.parameters;
    const content = elements.map((element: RootElement) => element.emit()).join('\n');
    return trim_every_line(content);
  }

}
