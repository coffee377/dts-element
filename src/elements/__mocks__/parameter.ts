export class Parameter {

  public parameters: any;

  constructor(parameters: any) {
    this.parameters = parameters;
  }

  public _emit(): string {
    return `[Parameter ${this.parameters.name}]`;
  }

}
