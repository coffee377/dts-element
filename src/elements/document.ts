import {Element} from '../element';
import {emit_elements} from '../helpers/emit-elements';
import {trim_every_line} from '../helpers/trim-every-line';
import {Stack} from '../stack';
import {Declaration} from './declaration';
import {ImportExport} from './import-export';
import {Reference} from './reference';

export interface IDocumentRequiredParameters {
  children: (Reference | Declaration | ImportExport)[];
}

// tslint:disable-next-line no-empty-interface
export interface IDocumentOptionalParameters {}

export class Document extends Element<IDocumentRequiredParameters, IDocumentOptionalParameters> {

  public get default_parameters(): IDocumentOptionalParameters {
    return {};
  }

  public _emit(stack: Stack): string {
    const {children} = this.parameters;
    const content = emit_elements(children, stack);
    return trim_every_line(content);
  }

}
