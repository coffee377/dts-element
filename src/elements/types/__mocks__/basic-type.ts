import {Type} from '../../type';

export class BasicType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
