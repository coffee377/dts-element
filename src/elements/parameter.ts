import {ParameterFlag} from '../collections';
import {any_type} from '../constants';
import {AnyElement, Element} from '../element';
import {emit_parameter_main} from '../helpers/emit-parameter-main';
import {emit_parameter_type} from '../helpers/emit-parameter-type';
import {AnyType} from './type';

export interface IParameterRequiredParameters {
  name: string;
}

export interface IParameterOptionalParameters {
  type: AnyType;
  flag: ParameterFlag;
}

export class Parameter extends Element<IParameterRequiredParameters, IParameterOptionalParameters> {

  public get default_parameters(): IParameterOptionalParameters {
    return {
      type: any_type,
      flag: null,
    };
  }

  public _emit(container: AnyElement | null): string {
    const {name, type, flag} = this.parameters;
    return `${emit_parameter_main(name, flag)}: ${emit_parameter_type(type, flag, container)}`;
  }

}
