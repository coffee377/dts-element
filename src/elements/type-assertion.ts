import {any_type} from '../constants';
import {Element} from '../element';
import {Stack} from '../stack';
import {Parameter} from './parameter';
import {Type} from './type';

export interface ITypeAssertionRequiredParameters {
  parameter: Parameter;
}

export interface ITypeAssertionOptionalParameters {
  type: Type;
}

export class TypeAssertion
    extends Element<ITypeAssertionRequiredParameters, ITypeAssertionOptionalParameters> {

  public get default_parameters(): ITypeAssertionOptionalParameters {
    return {
      type: any_type,
    };
  }

  public _emit(stack: Stack): string {
    const {parameter, type} = this.parameters;
    return `${parameter.parameters.name} is ${type.emit(stack)}`;
  }

}
