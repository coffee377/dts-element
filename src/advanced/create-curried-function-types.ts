import {FunctionDeclaration} from '../elements/declarations/function-declaration';
import {TypeDeclaration} from '../elements/declarations/type-declaration';
import {ObjectMember} from '../elements/members/object-member';
import {Parameter} from '../elements/parameter';
import {Type} from '../elements/type';
import {TypeAssertion} from '../elements/type-assertion';
import {FunctionType} from '../elements/types/function-type';
import {GenericType} from '../elements/types/generic-type';
import {ObjectType} from '../elements/types/object-type';
import {TypedType} from '../elements/types/typed-type';
import {left_pad} from '../utils/left-pad';

export interface ICreateCurriedFunctionTypesRequiredParameters {
  name: string;
  type: FunctionType;
}

export interface ICreateCurriedFunctionTypesOptionalParameters {
  placeholder: Type | null;
  generate_type_name(index: number): string;
  generate_placeholder_name(parameter_name: string): string;
}

export type ICreateCurriedFunctionTypesParameters =
  ICreateCurriedFunctionTypesRequiredParameters & Partial<ICreateCurriedFunctionTypesOptionalParameters>;

// tslint:disable:no-magic-numbers no-bitwise

/**
 * ```ts
 * type map = map_00;
 * type map_00 = {
 *   <T, U>(fn: (v: T) => U, array: T[]): map_11<T, U>;
 *   <T>(_fn: PH, array: T[]): map_01<T>;
 *   <T, U>(fn: (v: T) => U): map_10<T, U>;
 * };
 * type map_01<T> = {
 *   <U>(fn: (v: T) => U): map_11<T, U>;
 * };
 * type map_10<T, U> = {
 *   (array: T[]): map_11<T, U>;
 * };
 * type map_11<T, U> = U[];
 * ```
 */
export const create_curried_function_types = ({
      name,
      type,
      placeholder = null,
      generate_type_name = (index: number): string =>
        (`${name}_${left_pad(index.toString(2), type.parameters.parameters.length, '0')}`),
      generate_placeholder_name = (parameter_name: string): string => `_${parameter_name}`,
    }: ICreateCurriedFunctionTypesParameters): TypeDeclaration[] => {
  const curried_type_declarations = [...new Array(2 ** type.parameters.parameters.length)]
    .map((_value: any, index: number) => new TypeDeclaration({name: generate_type_name(index)}));
  const {parameters, generics} = type.parameters;
  const placeholder_parameters = (placeholder === null)
    ? null
    : parameters.map((parameter: Parameter): Parameter => new Parameter({
      name: generate_placeholder_name(parameter.parameters.name),
      type: placeholder,
    }));

  const return_type = type.parameters.return;
  if (return_type instanceof TypeAssertion) {
    throw new Error('AssertionType is not supported in create_curried_function_types');
  }

  curried_type_declarations.forEach((type_declaration: TypeDeclaration, index: number) => {

    const used_parameters: Parameter[] = [];
    const rest_parameters: Parameter[] = [];
    const rest_placeholders: Parameter[] = [];

    const is_parameter_used = (parameter_index: number): boolean =>
      (((2 ** (parameters.length - parameter_index - 1)) & ~index) === 0);

    parameters.forEach((parameter: Parameter, parameter_index: number) => {
      if (is_parameter_used(parameter_index)) {
        used_parameters.push(parameter);
      } else {
        rest_parameters.push(parameter);
        if (placeholder_parameters !== null) {
          rest_placeholders.push(placeholder_parameters[parameter_index]);
        }
      }
    });

    const used_generics = generics.filter((generic: GenericType) =>
      used_parameters.some((used_parameter: Parameter) => used_parameter.has(generic, true)));
    const rest_generics = generics.filter((generic: GenericType): boolean =>
      used_generics.indexOf(generic) === -1);

    type_declaration.parameters.generics.push(...used_generics);

    const curry_level = rest_parameters.length;
    const is_return_type_declaration = (index === curried_type_declarations.length - 1);

    if (is_return_type_declaration) {
      type_declaration.parameters.type = return_type;
    } else {
      const object_type = new ObjectType({members: []});
      type_declaration.parameters.type = object_type;
      // tslint:disable-next-line:only-arrow-functions
      (function recursive(
          current_parameters: Parameter[],
          current_used_generics: GenericType[],
          current_rest_generics: GenericType[]): void {
        if (current_parameters.length > 0 &&
            current_parameters[current_parameters.length - 1].parameters.type !== placeholder) {

          const rest_mask = current_parameters.reduce(
            (mask: number, parameter: Parameter) =>
              (parameter.parameters.type === placeholder)
                ? mask
                : mask | (1 << (parameters.length - parameters.indexOf(parameter) - 1)),
            index,
          );

          object_type.parameters.members.push(new ObjectMember({
            owned: new FunctionDeclaration({
              name: null,
              type: new FunctionType({
                generics: current_used_generics,
                parameters: current_parameters,
                return: new TypedType({
                  owned: curried_type_declarations[rest_mask],
                  generics: curried_type_declarations[rest_mask].parameters.generics,
                }),
              }),
            }),
          }));
        }
        const current_index = current_parameters.length;
        if (current_index < curry_level) {
          const current_parameter = rest_parameters[current_index];
          const current_placeholder = rest_placeholders[current_index];

          const current_parameter_rest_generics: GenericType[] = [];
          const current_parameter_used_generics: GenericType[] = [];

          current_rest_generics.forEach((generic: GenericType) => {
            if (current_parameter.has(generic, true)) {
              current_parameter_used_generics.push(generic);
            } else {
              current_parameter_rest_generics.push(generic);
            }
          });

          recursive(
            [...current_parameters, current_parameter],
            [...current_used_generics, ...current_parameter_used_generics],
            current_parameter_rest_generics,
          );
          if (placeholder !== null) {
            recursive(
              [...current_parameters, current_placeholder],
              current_used_generics,
              current_rest_generics,
            );
          }
        }
      })([], [], rest_generics);

      object_type.parameters.members.sort((member1: ObjectMember, member2: ObjectMember) => {
        const function_type_1: FunctionType = member1.parameters.owned.parameters.type;
        const function_type_2: FunctionType = member2.parameters.owned.parameters.type;
        return function_type_2.parameters.parameters.length - function_type_1.parameters.parameters.length;
      });
    }
  });

  curried_type_declarations.unshift(new TypeDeclaration({
    name,
    type: new TypedType({
      owned: curried_type_declarations[0],
    }),
  }));

  return curried_type_declarations;
};
