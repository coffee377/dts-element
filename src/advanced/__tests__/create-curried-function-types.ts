jest.disableAutomock();

import {any_type, number_type} from '../../constants';
import {Document} from '../../elements/document';
import {Parameter} from '../../elements/parameter';
import {TypeAssertion} from '../../elements/type-assertion';
import {ArrayType} from '../../elements/types/array-type';
import {BasicType} from '../../elements/types/basic-type';
import {FunctionType} from '../../elements/types/function-type';
import {GenericType} from '../../elements/types/generic-type';
import {UnionType} from '../../elements/types/union-type';
import {create_curried_function_types} from '../create-curried-function-types';

const name = 'adjust';
const generic_T = new GenericType({name: 'T'});
const generic_U = new GenericType({name: 'U'});
const function_type = new FunctionType({
  generics: [generic_T, generic_U],
  parameters: [
    new Parameter({
      name: 'fn',
      type: new FunctionType({
        parameters: [
          new Parameter({name: 'v', type: generic_T}),
        ],
        return: generic_U,
      }),
    }),
    new Parameter({name: 'index', type: number_type}),
    new Parameter({name: 'array', type: new ArrayType({owned: generic_T})}),
  ],
  return: new ArrayType({owned: new UnionType({types: [generic_T, generic_U]})}),
});

const parameter_value = new Parameter({
  name: 'value',
  type: any_type,
});
const function_type_assertion = new FunctionType({
  generics: [generic_T],
  parameters: [
    parameter_value,
    new Parameter({
      name: 'instance',
      type: generic_T,
    }),
  ],
  return: new TypeAssertion({
    parameter: parameter_value,
    type: generic_T,
  }),
});

it('should return correctly without placeholder', () => {
  const types = create_curried_function_types({name, type: function_type});
  const document = new Document({children: types});
  expect(document.emit()).toMatchSnapshot();
});

it('should return correctly with placeholder', () => {
  const placeholder = new BasicType({name: 'PH'});
  const types = create_curried_function_types({name, type: function_type, placeholder});
  const document = new Document({children: types});
  expect(document.emit()).toMatchSnapshot();
});

it('should return correctly with placeholder and selectable', () => {
  const placeholder = new BasicType({name: 'PH'});
  const types = create_curried_function_types({name, type: function_type, placeholder, selectable: true});
  const document = new Document({children: types});
  expect(document.emit()).toMatchSnapshot();
});

it('should throw error while using TypeAssertion', () => {
  const placeholder = new BasicType({name: 'PH'});
  // tslint:disable-next-line:typedef
  const create_types = () => create_curried_function_types({name, type: function_type_assertion, placeholder});
  expect(create_types).toThrowError();
});
