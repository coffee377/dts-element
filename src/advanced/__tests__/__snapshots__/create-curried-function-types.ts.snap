// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly with placeholder 1`] = `
"type adjust = adjust_000;
type adjust_000 = {
  <T, U>(fn: (v: T) => U, index: number, array: T[]): adjust_111<T, U>;
  <T, U>(fn: (v: T) => U, _index: PH, array: T[]): adjust_101<T, U>;
  <T>(_fn: PH, index: number, array: T[]): adjust_011<T>;
  <T>(_fn: PH, _index: PH, array: T[]): adjust_001<T>;
  <T, U>(fn: (v: T) => U, index: number): adjust_110<T, U>;
  (_fn: PH, index: number): adjust_010;
  <T, U>(fn: (v: T) => U): adjust_100<T, U>;
};
type adjust_001<T> = {
  <U>(fn: (v: T) => U, index: number): adjust_111<T, U>;
  (_fn: PH, index: number): adjust_011<T>;
  <U>(fn: (v: T) => U): adjust_101<T, U>;
};
type adjust_010 = {
  <T, U>(fn: (v: T) => U, array: T[]): adjust_111<T, U>;
  <T>(_fn: PH, array: T[]): adjust_011<T>;
  <T, U>(fn: (v: T) => U): adjust_110<T, U>;
};
type adjust_011<T> = {
  <U>(fn: (v: T) => U): adjust_111<T, U>;
};
type adjust_100<T, U> = {
  (index: number, array: T[]): adjust_111<T, U>;
  (_index: PH, array: T[]): adjust_101<T, U>;
  (index: number): adjust_110<T, U>;
};
type adjust_101<T, U> = {
  (index: number): adjust_111<T, U>;
};
type adjust_110<T, U> = {
  (array: T[]): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];"
`;

exports[`should return correctly without placeholder 1`] = `
"type adjust = adjust_000;
type adjust_000 = {
  <T, U>(fn: (v: T) => U, index: number, array: T[]): adjust_111<T, U>;
  <T, U>(fn: (v: T) => U, index: number): adjust_110<T, U>;
  <T, U>(fn: (v: T) => U): adjust_100<T, U>;
};
type adjust_001<T> = {
  <U>(fn: (v: T) => U, index: number): adjust_111<T, U>;
  <U>(fn: (v: T) => U): adjust_101<T, U>;
};
type adjust_010 = {
  <T, U>(fn: (v: T) => U, array: T[]): adjust_111<T, U>;
  <T, U>(fn: (v: T) => U): adjust_110<T, U>;
};
type adjust_011<T> = {
  <U>(fn: (v: T) => U): adjust_111<T, U>;
};
type adjust_100<T, U> = {
  (index: number, array: T[]): adjust_111<T, U>;
  (index: number): adjust_110<T, U>;
};
type adjust_101<T, U> = {
  (index: number): adjust_111<T, U>;
};
type adjust_110<T, U> = {
  (array: T[]): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];"
`;
